### PR TITLE
Move generic utility scripts to a common directory

### PR DESCRIPTION
**Pull request by MonicaisHer**
### Test
```
git clone https://github.com/MonicaisHer/edgex-checkbox-provider.git
cd edgex-checkbox-provider/data/jakarta/
git checkout move-utils-script
time (sudo ./run-all-tests-locally.sh)
```
Or you can test one single test. Here is an example:
```
time (sudo ./run-all-tests-locally.sh  -t test-self-refresh-services.sh)
```

### CODE REVIEW

**Comment by farshidtz** [comment]
If the function is specific to a release (e.g. check jakarta services), it better to keep it in the same directory. Generic ones such as install/refresh can be moved to a common location.

**Comment by MonicaisHer** [comment]
@farshidtz Thanks for your advice. This function `snap_check_jakarta_svcs` is not strictly tied to the Jakarta release. Jakarta and Ireland have the same function. I think for future releases, most of this function will be kept. So I chang `snap_check_jakarta_svcs` to `snap_check_svcs` and moved it to a common location. Is this sound good?

**Comment by farshidtz** [comment]
> @farshidtz Thanks for your advice. This function `snap_check_jakarta_svcs` is not strictly tied to the Jakarta release. Jakarta and Ireland have the same function. I think for future releases, most of this function will be kept. So I chang `snap_check_jakarta_svcs` to `snap_check_svcs` and moved it to a common location. Is this sound good?

This is the current situation but we can't make this assumption for future. The list of services or service names may change. Some services may be moved out to another snap.

I'd only add these to a common utils script:

- `snap_install`
- `snap_refresh`
- `get_snap_svc_status`
- `snap_remove`

And keep `snap_check_svcs` in the service directory's utils script. From there you can probably `source` the common utils.

You can go one step further and refactor `snap_check_svcs` and break it into 4 functions (check enabled, active, disabled, inactive). Those generic functions can then move to the common utils and called from the service-specific utils with appropriate input list.

**Comment by MonicaisHer** [comment]
@farshidtz  Thanks for your suggestion! I have updated them. Please have a look.

File: data/utils.sh [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+snap_remove()
+{
+    snap remove edgexfoundry || true
+}
+
+# wait for services to come online
+# NOTE: this may have to be significantly increased on arm64 or low RAM platforms
+# to accomodate time for everything to come online
+snap_wait_all_services_online()
```
Comment (farshidtz) [inline data/utils.sh]: This function is currently release-specific, especially the port checks in the end.

File: data/utils.sh [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        fi
+
+        if [[ "$core_data_status_code" == 200 ]] \
+            && [[ "$core_metadata_status_code" == 200 ]] \
+            && [[ "$core_command_status_code" == 200 ]] \
+            && [ -n "$(lsof -i -P -n -S 2 | grep 8000)" ] \
+            && [ -n "$(lsof -i -P -n -S 2 | grep 8200)" ] \
+            && [ -n "$(lsof -i -P -n -S 2 | grep 8500)" ] \
+            && [ -n "$(lsof -i -P -n -S 2 | grep 5432)" ] \
+            && [ -n "$(lsof -i -P -n -S 2 | grep 6379)" ]; then
```
Comment (farshidtz) [inline data/utils.sh]: Why not use the `snap_wait_port_status` function for these?

**Comment by farshidtz** [comment]
Please check if this is possible:

> And keep `snap_check_svcs` in the service directory's utils script. From there you can probably `source` the common utils.

File: data/utils.sh (hunk 1 above)
Comment (MonicaisHer) [reply data/utils.sh]: Thanks! I have removed it to jakarta/utils.sh now

File: data/utils.sh (hunk 2 above)
Comment (MonicaisHer) [reply data/utils.sh]: Updated. Thanks!

**Comment by MonicaisHer** [comment]
@farshidtz I have kept `snap_check_svcs` in the service directory's utils script now. Thanks!

File: data/jakarta/utils.sh [hunk 3]
```diff
@@ -5,125 +5,67 @@ if [ "$(id -u)" != "0" ]; then
...
-}
-
-snap_check_jakarta_svcs()
+snap_check_svcs()
 {
     if [ "$1" = "--notfatal" ]; then
         FATAL=0
     else
         FATAL=1
     fi
```
Comment (farshidtz) [inline data/jakarta/utils.sh]: This seems obsolete.

File: data/utils.sh [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        if [ "disabled" != "$svcStatus" ]; then
+            echo "service $svc has status \"$svcStatus\" but should be disabled"
+            if [ "$FATAL" = "1" ]; then
+                exit 1
+            fi
+        fi
+    done
+}
+
+# all the disabled services + the oneshot daemons
```
Comment (farshidtz) [inline data/utils.sh]: Misplaced comment. Please move to where this function is called.

File: data/jakarta/utils.sh (hunk 3 above)
Comment (farshidtz) [reply data/jakarta/utils.sh]: I found that it is used in the functions that are moved to the common utils. To make it clear, it is better to pass FATAL directly to the function.

Or maybe at least add a comment explaining that this variable is read by the sourced functions,

**Comment by farshidtz** [comment]
It looks good overall. I added a couple of minor comments.

File: data/jakarta/utils.sh (hunk 3 above)
Comment (MonicaisHer) [reply data/jakarta/utils.sh]: Thanks! I find that `--notfatal` for test-refresh-service is redundant. With `--notfatal`, if there are services in the wrong status, this test will still be passed. I suggest removing the `--notfatal` and `FATAL` variable.

File: data/jakarta/utils.sh [hunk 5]
```diff
@@ -5,125 +5,67 @@ if [ "$(id -u)" != "0" ]; then
...
+        "app-service-configurable kuiper device-virtual \
+        `#one-shot security services` \
+        security-proxy-setup \
+        security-secretstore-setup \
+        security-bootstrapper-redis \
+        security-consul-bootstrapper \
+        `#support services, system service` \
+        support-notifications \
+        support-scheduler \
+        sys-mgmt-agent"
```
Comment (farshidtz) [inline data/jakarta/utils.sh:68]: Writing these strings are really error prone. Can you try defining them as arrays?

Comment (farshidtz) [inline data/jakarta/utils.sh]: Writing these strings are really error-prone. Can you try defining them as arrays?

File: data/jakarta/utils.sh (hunk 3 above)
Comment (MonicaisHer) [reply data/jakarta/utils.sh]: Removed.

File: data/utils.sh (hunk 4 above)
Comment (MonicaisHer) [reply data/utils.sh]: Removed, thanks!

File: data/jakarta/utils.sh (hunk 5 above)
Comment (MonicaisHer) [reply data/jakarta/utils.sh]: Updated, thanks!

**Review by farshidtz — APPROVED** [review]
Looks good. Thanks a lot @MonicaisHer.

**Pull request merged**